File: py/layers.py
"""Layer stack + compositor engine — see spec/layers.md."""

from __future__ import annotations
import copy
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional
import numpy as np

MAX_LAYERS = 25
UNDO_DEPTH = 20


class BlendMode(Enum):
    Normal     = "normal"
    Multiply   = "multiply"
    Screen     = "screen"
    Overlay    = "overlay"
    Difference = "difference"


# ---------------------------------------------------------------------------
# sRGB <-> linear  (IEC 61966-2-1, same as plasma.py)
# ---------------------------------------------------------------------------

def srgb_to_linear(v: np.ndarray) -> np.ndarray:
    return np.where(v <= 0.04045, v / 12.92,
                    ((v + 0.055) / 1.055) ** 2.4)


def linear_to_srgb(v: np.ndarray) -> np.ndarray:
    return np.where(v <= 0.0031308, v * 12.92,
                    1.055 * np.power(np.maximum(v, 0.0), 1.0 / 2.4) - 0.055)


# ---------------------------------------------------------------------------
# Blend modes (linear, per-channel, top=a over bottom=b)
# ---------------------------------------------------------------------------

def _blend(mode: BlendMode, a: np.ndarray, b: np.ndarray) -> np.ndarray:
    if mode is BlendMode.Normal:
        return a
    if mode is BlendMode.Multiply:
        return a * b
    if mode is BlendMode.Screen:
        return 1.0 - (1.0 - a) * (1.0 - b)
    if mode is BlendMode.Overlay:
        return np.where(b < 0.5, 2.0 * a * b, 1.0 - 2.0 * (1.0 - a) * (1.0 - b))
    if mode is BlendMode.Difference:
        return np.abs(a - b)
    raise ValueError(f"Unknown blend mode {mode}")


# ---------------------------------------------------------------------------
# Layer
# ---------------------------------------------------------------------------

@dataclass
class Layer:
    """RGBA buffer in linear-light straight alpha."""
    rgba: np.ndarray          # shape (H, W, 4), float32, [0,1]
    mode: BlendMode = BlendMode.Normal
    opacity: float = 1.0
    visible: bool = True
    name: str = ""

    def copy(self) -> Layer:
        return Layer(self.rgba.copy(), self.mode, self.opacity, self.visible, self.name)


def make_layer(height: int, width: int, color=(0.0, 0.0, 0.0, 0.0),
               mode=BlendMode.Normal, opacity=1.0, visible=True, name="") -> Layer:
    buf = np.full((height, width, 4), color, dtype=np.float32)
    return Layer(buf, mode, opacity, visible, name)


# ---------------------------------------------------------------------------
# Composite
# ---------------------------------------------------------------------------

def composite(layers: list[Layer]) -> np.ndarray:
    """Flatten visible layers bottom->top; return (H,W,4) float32 linear."""
    if not layers:
        raise ValueError("empty layer list")
    H, W = layers[0].rgba.shape[:2]
    acc = np.zeros((H, W, 4), dtype=np.float64)  # (Cb_r, Cb_g, Cb_b, Ab)

    for layer in layers:
        if not layer.visible:
            continue
        src = layer.rgba.astype(np.float64)
        Cs = src[..., :3]
        As = src[..., 3:4] * float(layer.opacity)   # effective alpha

        Cb = acc[..., :3]
        Ab = acc[..., 3:4]

        blended = _blend(layer.mode, Cs, Cb)

        Aout = As + Ab * (1.0 - As)
        # Avoid div-by-zero; where Aout==0 result is 0
        safe = np.where(Aout > 0.0, Aout, 1.0)
        Cout = (blended * As + Cb * Ab * (1.0 - As)) / safe
        Cout = np.where(Aout > 0.0, Cout, 0.0)

        acc[..., :3] = Cout
        acc[..., 3:4] = Aout

    return np.clip(acc, 0.0, 1.0).astype(np.float32)


# ---------------------------------------------------------------------------
# Undo snapshot helpers
# ---------------------------------------------------------------------------

@dataclass
class _PixelSnap:
    layer_idx: int
    buf: np.ndarray

@dataclass
class _StructSnap:
    layers: list[Layer]   # deep copies


# ---------------------------------------------------------------------------
# Stack
# ---------------------------------------------------------------------------

class Stack:
    def __init__(self):
        self.layers: list[Layer] = []
        self._undo: list[_PixelSnap | _StructSnap] = []

    # -- undo ring --

    def _push_pixel_snap(self, idx: int):
        snap = _PixelSnap(idx, self.layers[idx].rgba.copy())
        self._undo.append(snap)
        if len(self._undo) > UNDO_DEPTH:
            self._undo.pop(0)

    def _push_struct_snap(self):
        snap = _StructSnap([l.copy() for l in self.layers])
        self._undo.append(snap)
        if len(self._undo) > UNDO_DEPTH:
            self._undo.pop(0)

    def undo(self) -> bool:
        if not self._undo:
            return False
        snap = self._undo.pop()
        if isinstance(snap, _PixelSnap):
            self.layers[snap.layer_idx].rgba = snap.buf
        else:
            self.layers = [l.copy() for l in snap.layers]
        return True

    # -- structural ops --

    def add(self, layer: Layer) -> Optional[str]:
        """Returns error string if at cap, else None."""
        if len(self.layers) >= MAX_LAYERS:
            return f"MAX_LAYERS ({MAX_LAYERS}) reached"
        self._push_struct_snap()
        self.layers.append(layer)
        return None

    def remove(self, idx: int):
        self._push_struct_snap()
        self.layers.pop(idx)

    def reorder(self, from_idx: int, to_idx: int):
        self._push_struct_snap()
        layer = self.layers.pop(from_idx)
        self.layers.insert(to_idx, layer)

    def flatten_visible(self):
        """Composite visible layers into one; preserve invisible in place."""
        self._push_struct_snap()
        visible_indices = [i for i, l in enumerate(self.layers) if l.visible]
        if not visible_indices:
            return
        merged_buf = composite(self.layers)
        merged = Layer(merged_buf, BlendMode.Normal, 1.0, True, "merged")
        insert_at = visible_indices[0]
        # Remove visible layers (iterate in reverse to preserve indices)
        for i in reversed(visible_indices):
            self.layers.pop(i)
        self.layers.insert(insert_at, merged)

    # -- pixel ops --

    def fill(self, idx: int, color: tuple[float, float, float, float]):
        self._push_pixel_snap(idx)
        self.layers[idx].rgba[...] = color

    def composite(self) -> np.ndarray:
        return composite(self.layers)

File: rs/src/layers.rs
//! Layer stack + compositor engine — see spec/layers.md.

pub const MAX_LAYERS: usize = 25;
pub const UNDO_DEPTH: usize = 20;

// ---------------------------------------------------------------------------
// sRGB <-> linear  (IEC 61966-2-1)
// ---------------------------------------------------------------------------

pub fn srgb_to_linear(v: f32) -> f32 {
    if v <= 0.04045 {
        v / 12.92
    } else {
        ((v + 0.055) / 1.055).powf(2.4)
    }
}

/// Encode linear float to sRGB; runs in f64 to match Python's ndarray f64 path.
pub fn linear_to_srgb_f64(v: f64) -> f64 {
    if v <= 0.0031308 {
        v * 12.92
    } else {
        1.055 * v.powf(1.0 / 2.4) - 0.055
    }
}

// ---------------------------------------------------------------------------
// Blend modes
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BlendMode {
    Normal,
    Multiply,
    Screen,
    Overlay,
    Difference,
}

/// Blend top channel `a` over bottom `b` — linear light, f64.
#[inline]
pub fn blend_channel(mode: BlendMode, a: f64, b: f64) -> f64 {
    match mode {
        BlendMode::Normal     => a,
        BlendMode::Multiply   => a * b,
        BlendMode::Screen     => 1.0 - (1.0 - a) * (1.0 - b),
        BlendMode::Overlay    => {
            if b < 0.5 { 2.0 * a * b } else { 1.0 - 2.0 * (1.0 - a) * (1.0 - b) }
        }
        BlendMode::Difference => (a - b).abs(),
    }
}

// ---------------------------------------------------------------------------
// Layer
// ---------------------------------------------------------------------------

#[derive(Clone)]
pub struct Layer {
    /// Linear-light straight-alpha RGBA, row-major, f32 per channel, [0,1].
    pub rgba: Vec<f32>,
    pub width: u32,
    pub height: u32,
    pub mode: BlendMode,
    pub opacity: f32,
    pub visible: bool,
    pub name: String,
}

impl Layer {
    pub fn new(width: u32, height: u32, color: [f32; 4]) -> Self {
        let n = (width * height * 4) as usize;
        let rgba = (0..n).map(|i| color[i % 4]).collect();
        Self {
            rgba,
            width,
            height,
            mode: BlendMode::Normal,
            opacity: 1.0,
            visible: true,
            name: String::new(),
        }
    }

    pub fn fill(&mut self, color: [f32; 4]) {
        for chunk in self.rgba.chunks_exact_mut(4) {
            chunk.copy_from_slice(&color);
        }
    }
}

// ---------------------------------------------------------------------------
// Composite  — accumulates in f64 throughout, stores result as f32.
// This matches Python's: acc in float64, .astype(np.float32) at the end.
// ---------------------------------------------------------------------------

pub fn composite(layers: &[Layer]) -> Layer {
    let first = layers.first().expect("empty layer list");
    let (w, h) = (first.width, first.height);
    let n = (w * h) as usize;

    // Four f64 channels: R G B A
    let mut acc_r = vec![0.0f64; n];
    let mut acc_g = vec![0.0f64; n];
    let mut acc_b = vec![0.0f64; n];
    let mut acc_a = vec![0.0f64; n];

    for layer in layers {
        if !layer.visible {
            continue;
        }
        let opacity = layer.opacity as f64;
        let mode    = layer.mode;

        for idx in 0..n {
            let i = idx * 4;
            let cs_r = layer.rgba[i    ] as f64;
            let cs_g = layer.rgba[i + 1] as f64;
            let cs_b = layer.rgba[i + 2] as f64;
            let cs_a = layer.rgba[i + 3] as f64;

            let as_ = cs_a * opacity;
            let ab  = acc_a[idx];
            let aout = as_ + ab * (1.0 - as_);

            if aout > 0.0 {
                let inv = 1.0 - as_;
                acc_r[idx] = (blend_channel(mode, cs_r, acc_r[idx]) * as_
                              + acc_r[idx] * ab * inv) / aout;
                acc_g[idx] = (blend_channel(mode, cs_g, acc_g[idx]) * as_
                              + acc_g[idx] * ab * inv) / aout;
                acc_b[idx] = (blend_channel(mode, cs_b, acc_b[idx]) * as_
                              + acc_b[idx] * ab * inv) / aout;
                acc_a[idx] = aout;
            }
            // if aout == 0: pixel stays zero (transparent black)
        }
    }

    // Clamp and convert to f32 — same as Python's np.clip(...).astype(np.float32)
    let mut rgba = vec![0.0f32; n * 4];
    for idx in 0..n {
        let i = idx * 4;
        rgba[i    ] = acc_r[idx].clamp(0.0, 1.0) as f32;
        rgba[i + 1] = acc_g[idx].clamp(0.0, 1.0) as f32;
        rgba[i + 2] = acc_b[idx].clamp(0.0, 1.0) as f32;
        rgba[i + 3] = acc_a[idx].clamp(0.0, 1.0) as f32;
    }

    Layer {
        rgba,
        width: w,
        height: h,
        mode: BlendMode::Normal,
        opacity: 1.0,
        visible: true,
        name: String::new(),
    }
}

// ---------------------------------------------------------------------------
// Undo snapshots
// ---------------------------------------------------------------------------

enum Snap {
    Pixel  { idx: usize, buf: Vec<f32> },
    Struct { layers: Vec<Layer> },
}

// ---------------------------------------------------------------------------
// Stack
// ---------------------------------------------------------------------------

pub struct Stack {
    pub layers: Vec<Layer>,
    undo: Vec<Snap>,
}

impl Stack {
    pub fn new() -> Self {
        Self { layers: Vec::new(), undo: Vec::new() }
    }

    /// Snapshot one layer's pixel buffer for undo without modifying it.
    /// Use before any destructive pixel op (punch, custom fill, etc.).
    pub fn snapshot_layer(&mut self, idx: usize) {
        self.push_pixel_snap(idx);
    }

    fn push_pixel_snap(&mut self, idx: usize) {
        let buf = self.layers[idx].rgba.clone();
        self.undo.push(Snap::Pixel { idx, buf });
        if self.undo.len() > UNDO_DEPTH {
            self.undo.remove(0);
        }
    }

    fn push_struct_snap(&mut self) {
        self.undo.push(Snap::Struct { layers: self.layers.clone() });
        if self.undo.len() > UNDO_DEPTH {
            self.undo.remove(0);
        }
    }

    pub fn add(&mut self, layer: Layer) -> Result<(), &'static str> {
        if self.layers.len() >= MAX_LAYERS {
            return Err("MAX_LAYERS reached");
        }
        self.push_struct_snap();
        self.layers.push(layer);
        Ok(())
    }

    pub fn remove(&mut self, idx: usize) {
        self.push_struct_snap();
        self.layers.remove(idx);
    }

    pub fn reorder(&mut self, from_idx: usize, to_idx: usize) {
        self.push_struct_snap();
        let layer = self.layers.remove(from_idx);
        self.layers.insert(to_idx, layer);
    }

    pub fn flatten_visible(&mut self) {
        self.push_struct_snap();
        let visible: Vec<usize> = self.layers.iter().enumerate()
            .filter(|(_, l)| l.visible)
            .map(|(i, _)| i)
            .collect();
        if visible.is_empty() {
            return;
        }
        let mut merged = composite(&self.layers);
        merged.mode    = BlendMode::Normal;
        merged.opacity = 1.0;
        merged.visible = true;
        merged.name    = "merged".into();
        let insert_at = visible[0];
        for &i in visible.iter().rev() {
            self.layers.remove(i);
        }
        self.layers.insert(insert_at, merged);
    }

    pub fn fill(&mut self, idx: usize, color: [f32; 4]) {
        self.push_pixel_snap(idx);
        self.layers[idx].fill(color);
    }

    pub fn undo(&mut self) -> bool {
        match self.undo.pop() {
            None => false,
            Some(Snap::Pixel { idx, buf }) => {
                self.layers[idx].rgba = buf;
                true
            }
            Some(Snap::Struct { layers }) => {
                self.layers = layers;
                true
            }
        }
    }

    pub fn composite(&self) -> Layer {
        composite(&self.layers)
    }
}
